### ISSUE TITLE
Unable to set up project on Windows

### ISSUE THREAD

**Issue by MarleyLP** [shamblett/C-NativeFunction#1]
I'm trying to follow the documentation on Medium while using Windows, I've installed WSL, https://learn.microsoft.com/en-us/windows/wsl/install, so that I could run the `sh build.sh` command, but I am still coming across the following error when I run said command:

```
add.c: In function ‘main’:
add.c:23:55: warning: format ‘%d’ expects a matching ‘int’ argument [-Wformat=]
   23 |         printf("Oops something is wrong the reult is %d\n\n");
      |                                                      ~^
      |                                                       |
      |                                                       int
writeln.c: In function ‘main’:
writeln.c:23:55: warning: format ‘%d’ expects a matching ‘int’ argument [-Wformat=]
   23 |         printf("Oops something is wrong the reult is %d\n\n");
      |                                                      ~^
      |                                                       |
      |                                                       int
sumstring.c: In function ‘main’:
sumstring.c:24:55: warning: format ‘%s’ expects a matching ‘char *’ argument [-Wformat=]
   24 |         printf("Oops something is wrong the reult is %s\n\n");
      |                                                      ~^
      |                                                       |
      |                                                       char *
```
      
I've also run dart add.dart in the dart directory, which gives me this error:

```
Opening.....
Unhandled exception:
Invalid argument(s): Failed to load dynamic library '../bin/debug/net6.0/linux-x64/publish/NativeLibrary.so': error code 126
#0      _open (dart:ffi-patch/ffi_dynamic_library_patch.dart:11:43)
#1      new DynamicLibrary.open (dart:ffi-patch/ffi_dynamic_library_patch.dart:22:12)
#2      main (file:///C:/Users/mlphelps/NativeLibrary/dart/add.dart:9:30)
#3      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#4      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:192:26)
```

### Dotnet Installation

I'm using dotnet 6.0.113, runtime 6.0.13, on my WSL, and dotnet 6.0.046, runtime 6.0.14, in my Windows Command Prompt.
My machine has had dotnet 3 and 7 on it recently but I got rid of them before attempting this.

### Using multiple terminals

I am running different commands in different terminals, because I was running into other errors when I tried to only use one.

I'm running the first couple commands in Windows Command Prompt because WSL would throw errors
```
dotnet new classlib -o NativeLibrary
dotnet add package Microsoft.DotNet.ILCompiler -v 7.0.0-*
dotnet publish /p:NativeLib=Shared /p:SelfContained=true -r win-x64
``` 
(The documentation has `linux-x64` instead of `win-x64` but I've changed it since I'm running in Windows Command Prompt)

I'm running `sh build.sh` in WSL because Windows Command Prompt can't run Bash scripts.

I'm running `dart add.dart` in my Android Studio terminal.

## Errors when only using WSL to run the commands

Error when running `dotnet new classlib -o NativeLibrary` in WSL
```
The template "Class Library" was created successfully.

Processing post-creation actions...
Running 'dotnet restore' on /mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj...
  Determining projects to restore...
/usr/lib/dotnet/sdk/6.0.113/NuGet.targets(130,5): error : Unable to obtain lock file access on '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' for operations on '/mnt/c/Users/mlphelps/NativeLibrary/obj/project.nuget.cache'. This may mean that a different user or administrator is holding this lock and that this process does not have permission to access it. If no other process is currently performing an operation on this file it may mean that an earlier NuGet process crashed and left an inaccessible lock file, in this case removing the file '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' will allow NuGet to continue. [/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj]
Restore failed.
Post action failed.
Manual instructions: Run 'dotnet restore'
```

Error when running `dotnet restore NativeLibrary.csproj` in WSL
```
  Determining projects to restore...
/usr/lib/dotnet/sdk/6.0.113/NuGet.targets(130,5): error : Unable to obtain lock file access on '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' for operations on '/mnt/c/Users/mlphelps/NativeLibrary/obj/project.nuget.cache'. This may mean that a different user or administrator is holding this lock and that this process does not have permission to access it. If no other process is currently performing an operation on this file it may mean that an earlier NuGet process crashed and left an inaccessible lock file, in this case removing the file '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' will allow NuGet to continue. [/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj]
```

Error when running `dotnet add package Microsoft.DotNet.ILCompiler -v 7.0.0-*` in WSL

```
Determining projects to restore...
  Writing /tmp/tmpB1qQjM.tmp
info : Adding PackageReference for package 'Microsoft.DotNet.ILCompiler' into project '/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj'.
info : Restoring packages for /mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj...
info :   GET https://api.nuget.org/v3-flatcontainer/microsoft.dotnet.ilcompiler/index.json
info :   OK https://api.nuget.org/v3-flatcontainer/microsoft.dotnet.ilcompiler/index.json 131ms
info :   GET https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/fe324174-469a-4296-a9cb-1fd885bb6f2b/nuget/v3/flat2/microsoft.dotnet.ilcompiler/index.json
info :   OK https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/fe324174-469a-4296-a9cb-1fd885bb6f2b/nuget/v3/flat2/microsoft.dotnet.ilcompiler/index.json 167ms
info : Package 'Microsoft.DotNet.ILCompiler' is compatible with all the specified frameworks in project '/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj'.
info : PackageReference for package 'Microsoft.DotNet.ILCompiler' version '7.0.0-*' added to file '/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj'.
info : Committing restore...
info : Generating MSBuild file /mnt/c/Users/mlphelps/NativeLibrary/obj/NativeLibrary.csproj.nuget.g.props.
info : Generating MSBuild file /mnt/c/Users/mlphelps/NativeLibrary/obj/NativeLibrary.csproj.nuget.g.targets.
info : Writing assets file to disk. Path: /mnt/c/Users/mlphelps/NativeLibrary/obj/project.assets.json
error: Unable to obtain lock file access on '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' for operations on '/mnt/c/Users/mlphelps/NativeLibrary/obj/project.nuget.cache'. This may mean that a different user or administrator is holding this lock and that this process does not have permission to access it. If no other process is currently performing an operation on this file it may mean that an earlier NuGet process crashed and left an inaccessible lock file, in this case removing the file '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' will allow NuGet to continue.


Usage: NuGet.CommandLine.XPlat.dll package add [options]

Options:
  -h|--help               Show help information
  --force-english-output  Forces the application to run using an invariant, English-based culture.
  --package               Id of the package to be added.
  --version               Version of the package to be added.
  -d|--dg-file            Path to the dependency graph file to be used to restore preview and compatibility check.
  -p|--project            Path to the project file.
  -f|--framework          Frameworks for which the package reference should be added.
  -n|--no-restore         Do not perform restore preview and compatibility check. The added package reference will be unconditional.
  -s|--source             Specifies NuGet package sources to use during the restore.
  --package-directory     Directory to restore packages in.
  --interactive           Allow the command to block and require manual action for operations like authentication.
  --prerelease            Allows prerelease packages to be installed.
  ```
  
Error when running `dotnet publish /p:NativeLib=Shared /p:SelfContained=true -r linux-x64 -c debug` in WSL (I tried both`linux-x64` and `win-x64`)
```
Microsoft (R) Build Engine version 17.0.1+b177f8fa7 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
/usr/lib/dotnet/sdk/6.0.113/NuGet.targets(130,5): error : Unable to obtain lock file access on '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' for operations on '/mnt/c/Users/mlphelps/NativeLibrary/obj/project.nuget.cache'. This may mean that a different user or administrator is holding this lock and that this process does not have permission to access it. If no other process is currently performing an operation on this file it may mean that an earlier NuGet process crashed and left an inaccessible lock file, in this case removing the file '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' will allow NuGet to continue. [/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj]
```

If I ignore these errors in WSL, when I run `sh build.sh` I get the same error I do as when I run the commands in Windows Command Prompt.

**Comment by shamblett** [comment]
I'm not a windows user, I develop on Linux however I do have a windows machine with WSL installed, I'll have  a look at this shortly.

Presumably the warnings in your add.c are not stopping the object file being created are they?

**Comment by MarleyLP** [comment]
Thank you.

No, the warnings don't stop the creation of the .exe or .ddl files.

**Comment by shamblett** [comment]
OK, this works for me OK on Win10 under WSL, same output as you on the compile step, I also had to add '-ldl' to the gcc commands in ./build.sh.

Looking at your errors above, this one -
```
/usr/lib/dotnet/sdk/6.0.113/NuGet.targets(130,5): error : Unable to obtain lock file access on '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' for operations on '/mnt/c/Users/mlphelps/NativeLibrary/obj/project.nuget.cache'. This may mean that a different user or administrator is holding this lock and that this process does not have permission to access it. If no other process is currently performing an operation on this file it may mean that an earlier NuGet process crashed and left an inaccessible lock file, in this case removing the file '/tmp/NuGetScratch/lock/ec496004be94eb9f5f3f109ca97cf7a128ce9ef7' will allow NuGet to continue. [/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj]
```
Is complaining about a lock file, this may be your multiple windows interfering with each other, its not the instructions per se.

Ive not tried it on native windows but in thepry you should just be able to change the '-r linux-x64' to 'win-x64' to get a dll, then you will have to edit the dart files to change -
```
 DynamicLibrary.open('../bin/debug/net6.0/linux-x64/publish/NativeLibrary.so');
```
to the dll that has been generated, something like -
```
 DynamicLibrary.open('../bin/debug/net6.0/win-x64/publish/NativeLibrary.dll');
```

you need to check the paths.

**Comment by MarleyLP** [comment]
Thank you, I just had to delete the lock files like the error message suggested, I had my blinkers on, as I was mainly focused on running it in native windows, but I think I'll stick to using purely WSL for now.

Unfortunately I still have errors, when running `dotnet publish /p:NativeLib=Shared /p:SelfContained=true -r linux-x64 -c debug` I get:

```
Microsoft (R) Build Engine version 17.0.1+b177f8fa7 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
/usr/lib/dotnet/sdk/6.0.113/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(1110,5): warning NETSDK1179: One of '--self-contained' or '--no-self-contained' options are required when '--runtime' is used. [/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj]
  NativeLibrary -> /mnt/c/Users/mlphelps/NativeLibrary/bin/debug/net6.0/linux-x64/NativeLibrary.dll
  /usr/bin/ld: cannot find -lstdc++: No such file or directory
collect2 : error : ld returned 1 exit status [/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj]
/home/marley/.nuget/packages/microsoft.dotnet.ilcompiler/7.0.0/build/Microsoft.NETCore.Native.targets(346,5): error MSB3073: The command ""gcc" "obj/debug/net6.0/linux-x64/native/NativeLibrary.o" -o "bin/debug/net6.0/linux-x64/native/NativeLibrary.so" -Wl,--version-script=obj/debug/net6.0/linux-x64/native/NativeLibrary.exports /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/sdk/libbootstrapperdll.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/sdk/libRuntime.WorkstationGC.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.Native.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.Globalization.Native.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.IO.Compression.Native.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.Net.Security.Native.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.Security.Cryptography.Native.OpenSsl.a -g -Wl,-rpath,'$ORIGIN' -Wl,--build-id=sha1 -Wl,--as-needed -Wl,-e0x0 -pthread -lstdc++ -ldl -lm -lz -lrt -shared -Wl,-z,relro -Wl,-z,now -Wl,--require-defined,NativeAOT_StaticInitialization -Wl,--discard-all -Wl,--gc-sections" exited with code 1. [/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj]
```

EDIT:

And when I run `dotnet publish /p:NativeLib=Shared /p:SelfContained=true -r linux-x64 -c debug --self-contained -verbosity:diag` it gives me this extra information:

```
      ...
      ...
      628 ms  Exec                                       3 calls
      664 ms  MSBuild                                    7 calls

Build FAILED.

                   "/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj" (Publish target) (1:7) ->
                   (LinkNative target) ->
                     collect2 : error : ld returned 1 exit status [/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj]
                     /home/marley/.nuget/packages/microsoft.dotnet.ilcompiler/7.0.0/build/Microsoft.NETCore.Native.targets(346,5): error MSB3073: The command ""gcc" "obj/debug/net6.0/linux-x64/native/NativeLibrary.o" -o "bin/debug/net6.0/linux-x64/native/NativeLibrary.so" -Wl,--version-script=obj/debug/net6.0/linux-x64/native/NativeLibrary.exports /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/sdk/libbootstrapperdll.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/sdk/libRuntime.WorkstationGC.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.Native.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.Globalization.Native.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.IO.Compression.Native.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.Net.Security.Native.a /home/marley/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/7.0.0/framework/libSystem.Security.Cryptography.Native.OpenSsl.a -g -Wl,-rpath,'$ORIGIN' -Wl,--build-id=sha1 -Wl,--as-needed -Wl,-e0x0 -pthread -lstdc++ -ldl -lm -lz -lrt -shared -Wl,-z,relro -Wl,-z,now -Wl,--require-defined,NativeAOT_StaticInitialization -Wl,--discard-all -Wl,--gc-sections" exited with code 1. [/mnt/c/Users/mlphelps/NativeLibrary/NativeLibrary.csproj]

    0 Warning(s)
    2 Error(s)
    ```

**Comment by shamblett** [comment]
There's something wrong with your build environment, for some reason, this -
```
-lstdc++: No such file or directory
```
is saying your libstdc++ library can't be found and your gcc command is exiting with an an exit code of 1.

Make sure you have installed a full C/C++ development environment on your target.

**Comment by MarleyLP** [comment]
Sorry, I'm honestly not sure if I have a full C/C++ development environment on my WSL.

I have gcc, g++ and c++ 11.3.0, gdb 12.1, cmake 3.22.1 valgrind 3.18.1, and emacs 27.1

Also now I'm getting `/usr/bin/ld: cannot find -lz: No such file or directory` instead.

**Comment by shamblett** [comment]
This means you don't have zlib installed, the best way to do this is to google these as they come up and install the required library, the 'lz' above translates to 'libz.so' for instance, the 'l' directive assumes the library starts with 'lib'.

**Comment by MarleyLP** [comment]
Thanks, sorry, I'm not used to Linux syntax, I installed zlib1g-dev (after installing zlib1g didn't fix anything) and now I can publish without errors.

Unfortunately I still have the original error when running `sh build.sh`:

```
add.c: In function ‘main’:
add.c:23:55: warning: format ‘%d’ expects a matching ‘int’ argument [-Wformat=]
   23 |         printf("Oops something is wrong the reult is %d\n\n");
      |                                                      ~^
      |                                                       |
      |                                                       int
/usr/bin/ld: cannot find add: No such file or directory
collect2: error: ld returned 1 exit status
writeln.c: In function ‘main’:
writeln.c:23:55: warning: format ‘%d’ expects a matching ‘int’ argument [-Wformat=]
   23 |         printf("Oops something is wrong the reult is %d\n\n");
      |                                                      ~^
      |                                                       |
      |                                                       int
/usr/bin/ld: cannot find writeln: No such file or directory
collect2: error: ld returned 1 exit status
sumstring.c: In function ‘main’:
sumstring.c:24:55: warning: format ‘%s’ expects a matching ‘char *’ argument [-Wformat=]
   24 |         printf("Oops something is wrong the reult is %s\n\n");
      |                                                      ~^
      |                                                       |
      |                                                       char *
/usr/bin/ld: cannot find sumstring: No such file or directory
collect2: error: ld returned 1 exit status
```

But now with added `/usr/bin/ld: cannot find file: No such file or directory` errors, after adding `-ldl`  to the build.sh file as you suggested:

```
#!/bin/bash
#
# C example builder
#
rm add writeln sumstring 2> /dev/null
gcc add.c -o -ldl add 
gcc writeln.c -o -ldl writeln
gcc sumstring.c -o -ldl sumstring
```

I tried getting the compiler to search the path by using `-L/mnt/c/Users/mlphelps/NativeLibrary/c` but it didn't change anything.

**Comment by shamblett** [comment]
OK, are you producing the object files

**Comment by MarleyLP** [comment]
No I'm not, I do if I remove the `-ldl`'s from the `build.sh`, or move them to anywhere else on the lines (this also gets rid of the `/usr/bin/ld: cannot find file: No such file or directory` but I assume only because `-ldl` just doesn't do anything in the wrong place).

**Comment by shamblett** [comment]
From the docs -
```
ibdl.so. This file is a dynamic library that contains functions like dlopen() or dlsym() : These functions allow a program to "dynamically" load dynamic libraries - this means the program can call a function to load a dynamic library
```
So if you remove this I'm assuming the code won't link. What distro are you using in WSL? I used Ubuntu 20.04, I did find however I didn't need to do this on my main Linux dev box which is Fedora based.

**Comment by MarleyLP** [comment]
`lsb_release -a` returns:
```
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.1 LTS
Release:        22.04
Codename:       jammy
```

**Comment by shamblett** [comment]
Have you checked that libdl.so is installed

**Comment by MarleyLP** [comment]
Yeah I have `libdl.so.2` in my `/usr/lib/x86_64-linux-gnu` directory

**Comment by shamblett** [comment]
Same place as mine, so looks as though there is something wrong with your set up somewhere.

**Comment by MarleyLP** [comment]
Hmm, I'll keep looking into it.

Thank you for all your help and patience, it is greatly appreciated :)

**Comment by MarleyLP** [comment]
I resolved the issue!

I just wasn't reading the error message correctly, it was a problem with the c files, not `build.sh`.

I was able to create the object files by removing the `-ldl`s from `build.sh` and by adding `, result` into add.c, sumstring.c, and writeln.c, changing `printf("Oops something is wrong the reult is %d\n\n");` to `printf("Oops something is wrong the reult is %d\n\n", result);` in add.c and writeln.c, and  changing `printf("Oops something is wrong the reult is %s\n\n");` to `printf("Oops something is wrong the reult is %s\n\n", result);` in sumstring.c.

And I was able to run the dart files simply by installing dart on my WSL instead of trying to use Android Studio.

Thanks again for all the help, and for the project as a whole :)

Edit: 

And since I don't actually need to run `sh build.sh` to import the C# to Dart I can use Windows command line instead of WSL,
(By running `dotnet publish /p:NativeLib=Shared /p:SelfContained=true -r win-x64 -c debug` instead of `dotnet publish /p:NativeLib=Shared /p:SelfContained=true -r linux-x64 -c debug` and by changing the path in the dart files to `final lib = DynamicLibrary.open('../bin/debug/net6.0/win-x64/publish/NativeLibrary.dll');` instead of `final lib = DynamicLibrary.open('../bin/debug/net6.0/linux-x64/publish/NativeLibrary.so');`, as you had previously mentioned)
